hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/upgrade/TestDatanodeUpgradeToSchemaV4.java

```diff
@@ -0,0 +1,383 @@
+        tempFolder.toPath().resolve("export-testContainerSchemaV4.tar")).toFile();
+    TarContainerPacker packer = new TarContainerPacker(NO_COMPRESSION);
+
+    //export the container
+    try (FileOutputStream fos = new FileOutputStream(folderToExport)) {
+      container.exportContainerData(fos, packer);
+    }
+
+    //delete the original one
+    KeyValueContainerUtil.removeContainer(container.getContainerData(), conf);
+    container.delete();
+    assertFalse(new File(container.getContainerData().getContainerPath()).exists());
+    if (schemaV3Enabled) {
+      assertTrue(container.getContainerData().getDbFile().exists());
+    }
+
+    //create a new one
+    KeyValueContainerData oldContainerData = container.getContainerData();
+    KeyValueContainerData newContainerData =
+        new KeyValueContainerData(containerID1,
+            oldContainerData.getLayoutVersion(),
+            oldContainerData.getMaxSize(), pipeline.getId().getId().toString(),
+            dsm.getDatanodeDetails().getUuidString());
+    newContainerData.setSchemaVersion(oldContainerData.getSchemaVersion());
+    KeyValueContainer newContainer = new KeyValueContainer(newContainerData, conf);
+    newContainer.populatePathFields(scmTestMock.getClusterId(), oldContainerData.getVolume());
+
+    // verify yaml file checksum
+    try (FileInputStream fis = new FileInputStream(folderToExport)) {
+      byte[] containerDescriptorYaml = packer.unpackContainerDescriptor(fis);
+      KeyValueContainerData data = (KeyValueContainerData) ContainerDataYaml
+          .readContainer(containerDescriptorYaml);
+      ContainerUtils.verifyChecksum(data, conf);
+    }
+
+    // sleep 1s to make sure creationTime will change
+    Thread.sleep(1000);
+    try (FileInputStream fis = new FileInputStream(folderToExport)) {
+      newContainer.importContainerData(fis, packer);
+    }
+
+    assertEquals(newContainerData.getContainerDBType(), oldContainerData.getContainerDBType());
+    assertEquals(newContainerData.getState(), oldContainerData.getState());
+    assertEquals(newContainerData.getBlockCount(), oldContainerData.getBlockCount());
+    assertEquals(newContainerData.getLayoutVersion(), oldContainerData.getLayoutVersion());
+    assertEquals(newContainerData.getMaxSize(), oldContainerData.getMaxSize());
+    assertEquals(newContainerData.getBytesUsed(), oldContainerData.getBytesUsed());
+    assertEquals(newContainerData.getMetadataPath(), oldContainerData.getMetadataPath());
+    assertEquals(newContainerData.getChunksPath(), oldContainerData.getChunksPath());
+    assertEquals(newContainerData.getContainerPath(), oldContainerData.getContainerPath());
+    assertTrue(new File(newContainerData.getContainerPath()).exists());
+    assertTrue(new File(newContainerData.getChunksPath()).exists());
+    assertTrue(new File(newContainerData.getMetadataPath()).exists());
+    if (schemaV3Enabled) {
+      assertTrue(newContainerData.getDbFile().exists());
```

Comment:
nit: `assertThat(<some File>).exists();` provides better error message if the assertion fails.